torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -329,6 +330,21 @@ def set_post_optim_event(self, event: torch.cuda.Event) -> None:
         """
         self._get_fsdp_state()._state_ctx.post_optim_event = event
 
+    def set_reduce_scatter_divide_factor(self, factor: float) -> None:
+        """
+        Sets a custom divide factor for the reduce-scatter. This becomes a
+        custom reduce op using NCCL's PreMulSum, which allows multiplying by
+        the factor before reduction.
+
+        Args:
+            factor (float): Custom divide factor.
+        """
+        state = self._get_fsdp_state()
+        if (fsdp_param_group := state._fsdp_param_group) is not None:
+            mul_factor = 1.0 / float(factor)
+            reduce_op = _make_nccl_premul_sum(mul_factor)
```

Comment:
Nice! A few questions:
- This doesn't yet work with compile right?
- Is there any case where we want to favor `_div_if_needed` over `premul_sum` when nccl is used?

Reply:
I am not sure about compile (have not tried).

I think we always prefer `premul_sum` for the pre-divide over `_div_if_needed`. I just did not know about this feature until I wrote this PR (which is for some internal use case exploration) 😄 .

Reply:
I updated the PR to get rid of `_div_if_needed` before reduction for fp16 path.

Reply:
Nevermind. MTPG does not support premulsum, and the Python plumbing is not there yet to support this. I will remove it from this PR for now.